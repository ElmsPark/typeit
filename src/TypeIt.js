import Instance from "./Instance";
import allHaveStatus from "./helpers/allHaveStatus";
import { generateHash } from "./utilities";
import toArrayOfNodes from "./helpers/toArrayOfNodes";

export default class TypeIt {
  constructor(element, options) {
    this.instances = toArrayOfNodes(element).map(el => {
      return new Instance({
        element: el,
        id: generateHash(),
        options,
        typeit: this
      });
    });
  }

  /**
   * Push a specific action directly into the queue of each instance.
   * If an instance has already completed, trigger the queeu again.
   *
   * @param {string} function
   * @param {*} argument
   * @
   */
  queueUp(action, argument = null, numberOfTimesToCopy = 1) {
    this.instances.forEach(instance => {
      let isIndependentFunction = typeof action !== "string";

      /**
       * If action is a string, bind it to instance.
       * Otherwise, leave it on its own.
       */
      let toFire = isIndependentFunction ? action : instance[action];

      let toPassAsArguments = isIndependentFunction ? this : argument;

      for (let i = 0; i < numberOfTimesToCopy; i++) {
        instance.queue.add([toFire, toPassAsArguments]);
      }
    });
  }

  is(status) {
    return allHaveStatus(this.instances, status, true);
  }

  freeze() {
    this.instances.forEach(instance => {
      instance.status.frozen = true;
    });
  }

  unfreeze() {
    this.instances.forEach(instance => {
      if (!instance.status.frozen) return;
      instance.status.frozen = false;
      instance.fire();
    });
  }

  /**
   * If used after typing has started, will append strings to the end of the existing queue. If used when typing is paused, will restart it.
   *
   * @param  {string} string The string to be typed.
   * @return {object} TypeIt instance
   */
  type(string = "") {
    this.instances.forEach(instance => instance.queueString(string));
    return this;
  }

  /**
   * If null is passed, will delete whatever's currently in the element.
   *
   * @param  { number } numCharacters Number of characters to delete.
   * @return { TypeIt }
   */
  delete(numberOfCharactersToDelete = null) {
    this.queueUp(
      'delete',
      numberOfCharactersToDelete === null, //-- Maybe delete all.
      numberOfCharactersToDelete === null ? 1 : numberOfCharactersToDelete
    );

    return this;
  }

  //-- works!
  pause(ms = null) {
    this.queueUp("pause", ms);
    return this;
  }

  break() {
    this.queueUp("insert", "<br>");
    return this;
  }

  options(options) {
    this.queueUp("setOptions", options);
    return this;
  }

  destroy(removeCursor = true) {
    this.instances.forEach(instance => {
      instance.timeouts.forEach(timeout => {
        clearTimeout(timeout);
      });

      instance.timeouts = [];

      if (removeCursor && instance.opts.cursor) {
        instance.$eWrapper.removeChild(
          instance.$eWrapper.querySelector(".ti-cursor")
        );
      }

      instance.status.destroyed = true;
    });
  }

  hook(func) {
    this.queueUp(func);
    return this;
  }

  /**
   * Reset each instance like it's brand new.
   */
  reset() {
    this.instances = this.instances.map(instance => {
      return instance.reset();
    });
    return this;
  }

  go() {
    this.instances.forEach(instance => {
      instance.init();
    });
    return this;
  }
}
