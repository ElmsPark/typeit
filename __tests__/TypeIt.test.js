import TypeIt from "../src/TypeIt.js";

let instance;

beforeEach(() => {
  document.body.innerHTML = `
    <div>
      <span id="element"></span>
    </div>
  `;

  instance = new TypeIt("#element");
});

test("Initial queue only contains startDelay pause.", () => {
  expect(instance.instances[0].queue).toMatchSnapshot();
});

test("Returns an object with expected properties.", () => {
  expect(Object.keys(instance).sort()).toEqual(["instances"]);
});

test("Defines hard-coded string correctly.", () => {
  jest.useFakeTimers();

  document.body.innerHTML = `
    <div>
      <span id="element">Hard-coded string.</span>
    </div>
  `;

  let instance = new TypeIt("#element", {
    strings: ["My string."]
  });

  expect(instance.instances[0].opts.strings).toEqual([
    "Hard-coded string.",
    "My string."
  ]);
});

test("Will not begin until explicitly called.", () => {
  jest.useFakeTimers();

  document.body.innerHTML = `<div>'
      <span id="element"></span>
    </div>`;

  const instance = new TypeIt("#element", {
    strings: "hello!"
  });

  jest.runAllTimers();

  expect(instance.is("started")).toBe(false);

  instance.go();

  expect(instance.is("started")).toBe(true);
});

test("Clears out remnants of previous instances correctly.", () => {
  jest.useFakeTimers();

  document.body.innerHTML = `
    <div>
      <span id="element">
        <span style="display:inline;position:relative;font:inherit;color:inherit;" class="ti-wrapper">Previous string.</span>
      </span>
    </div>
  `;

  let instance = new TypeIt("#element", {
    strings: "My string."
  });

  expect(
    !instance.instances[0].opts.strings[0].includes("ti-container")
  ).toEqual(true);
});

test("Typing doesn't end with a break tag.", () => {
  jest.useFakeTimers();

  document.body.innerHTML = `<div>'
    <span id="element"></span>
  </div>`;

  const instance = new TypeIt("#element", {
    strings: ["One string.", "Two string", "Three string."]
  }).go();

  jest.runAllTimers();

  expect(instance.instances[0].$eContainer.innerHTML.endsWith("<br>")).not.toBe(
    true
  );
});

describe("reset()", () => {
  test("Successfully resets when called.", () => {
    jest.useFakeTimers();

    document.body.innerHTML = `<div>'
        <span id="element"></span>
      </div>`;

    const instance = new TypeIt("#element", {
      strings: "This is my string!"
    }).go();

    jest.runAllTimers();

    instance.destroy();

    expect(instance.is("destroyed")).toBe(true);

    instance.reset();

    //-- Ensure the arguments that define these properties were passed.
    expect(instance.instances[0].$e).not.toBe(undefined);
    expect(instance.instances[0].id).not.toBe(undefined);
    expect(instance.instances[0].opts).not.toBe(undefined);
    expect(instance.instances[0].typeit).not.toBe(undefined);
    expect(instance.instances).toHaveLength(1);
    expect(instance.is("complete")).toBe(false);
    expect(instance.is("destroyed")).toBe(false);
  });
});

describe("destroy()", () => {
  test("Destroys instances successfully.", () => {
    jest.useFakeTimers();

    document.body.innerHTML = `<div>'
      <span id="element"></span>
    </div>`;

    const instance = new TypeIt("#element", {
      strings: "This is my string!"
    }).go();

    jest.runAllTimers();

    expect(instance.is("destroyed")).toBe(false);

    instance.destroy();

    expect(instance.instances[0].timeouts).toEqual([]);
    expect(document.body.querySelector(".ti-cursor")).toEqual(null);
  });
});

describe("queueUp()", () => {
  test("It should add function name by string w/ correct arguments.", () => {
    instance.queueUp("type", "someArg");
    expect(instance.instances[0].queue.waiting).toMatchSnapshot();
  });

  test("If invalid function name is passed, will save as undefined", () => {
    instance.queueUp("thisDoesNotExist", "someArg");
    expect(instance.instances[0].queue.waiting).toMatchSnapshot();
  });

  test("Correctly enqueues anonymous function.", () => {
    instance.queueUp(function() {
      return true;
    });

    let funcAsString = instance.instances[0].queue.waiting[1][0].toString();
    let argument = instance.instances[0].queue.waiting[1][1];

    expect(funcAsString).toMatchSnapshot();
    expect(argument).toBeInstanceOf(TypeIt);
  });

  test("It duplicates the action a certain number of times per instance, if specified.", () => {
    //-- Add 'type' to queue three times.
    instance.queueUp("type", 1, 3);

    expect(instance.instances[0].queue.waiting).toMatchSnapshot();

    instance.queueUp("delete", 1);

    expect(instance.instances[0].queue.waiting).toMatchSnapshot();
  });
});
