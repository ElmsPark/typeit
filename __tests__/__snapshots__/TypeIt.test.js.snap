// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queueUp() Correctly enqueues anonymous function. 1`] = `
"function () {
      return true;
    }"
`;

exports[`queueUp() If invalid function name is passed, will save as undefined 1`] = `
Array [
  Array [
    undefined,
    "someArg",
  ],
]
`;

exports[`queueUp() It duplicates the action a certain number of times per instance, if specified. 1`] = `
Array [
  Array [
    [Function],
    1,
  ],
  Array [
    [Function],
    1,
  ],
  Array [
    [Function],
    1,
  ],
]
`;

exports[`queueUp() It duplicates the action a certain number of times per instance, if specified. 2`] = `
Array [
  Array [
    [Function],
    1,
  ],
  Array [
    [Function],
    1,
  ],
  Array [
    [Function],
    1,
  ],
  Array [
    [Function],
    1,
  ],
]
`;

exports[`queueUp() It should add function name by string w/ correct arguments. 1`] = `
Array [
  Array [
    [Function],
    "someArg",
  ],
]
`;
